Add new logic to enable stored checkpoint weights to be copied to new history dimensions

scottcha opened the pull request:
Here is a possible fix for #35.  Currently the new logic attempts to copy the provided weights into any new history dimensions which get allocated by setting max_history_size > 2.

scottcha commented:
@microsoft-github-policy-service agree

wesselb reviewed (changes requested):
Hey @scottcha! Thanks for opening a PR!! I think this is useful feature. :) A few comments:

I think we should not allow the user to decrease the value of `max_history_size`, as that would very likely break performance. Could you add an assertion for this?

In addition, currently this only works for the pretrained checkpoint, not any checkpoint saved after that. Would you be able to separate the checkpoint-adaptation logic from the `max_history_size`-adaptation logic, so this would also work for checkpoints saved after that? We might at some point release additional models to which the checkpoint-adaptation logic does not apply, but for which you might want to change `max_history_size`.

scottcha commented:
@wesselb that sounds like good suggestions.  I'll update the PR addressing those.  It might take a couple days based on my availability but its no problem.

scottcha commented:
PR is updated with the requested changes. 
1. I scoped the changes to just the suggestions provided and limited to the history size adaptation.  
2. I might recommend that the checkpoint weight adaptation is also refactored to its own method helping make it more testable and overridable.  Let me know if you want to see this and I can provide a separate PR scoped to that change.
3. I provided a new test file for this method and added the relevant tests.
4. Beyond the unit tests provided I've also validated the functional cases here:
- AuroraSmall loads with provided small checkpoint with max_history_size=2 (default)
- AuroraSmall loads with provided small checkpoint with max_history_size=4
- AuroraSmall can save new checkpoint after loading with max_historysize=4 and that can be reloaded after being serialized
- AuroraSmall 2 step rollout continues work for 2t after loading checkpoints above
Also note that one AuroraSmall test case (test_aurora_small) seems to be failing on your main branch and it seems to be unrelated to my changes here.  All other tests are passing.
Let me know if you have any other feedback.
thanks!

wesselb commented on aurora/model/aurora.py:
```diff
@@ -316,8 +316,42 @@ def load_checkpoint_local(self, path: str, strict: bool = True) -> None:
+                    into model with max_history_size {self.max_history_size} for weight {name}"   
+            
+                # Initialize the new weight tensor with zeros
+                new_weight = torch.zeros((weight.shape[0], 1, self.max_history_size, weight.shape[3], weight.shape[4]))
```
Should we set the device and dtype equal to the existing weight? It should usually be CPU and float32, the the defaults, but that might not universally be the case.

scottcha replied:
Done.

wesselb commented on aurora/model/aurora.py:
```diff
@@ -316,8 +316,42 @@ def load_checkpoint_local(self, path: str, strict: bool = True) -> None:
+                new_weight = torch.zeros((weight.shape[0], 1, self.max_history_size, weight.shape[3], weight.shape[4]))
+                # Copy the existing weights to the new tensor my duplicating the histories provided in to any new history dimensions
+                for j in range(new_weight.shape[2]):
+                    new_weight[:, :, j, :, :] = weight[:, :, j % weight.shape[2], :, :]
```
I'm wondering about the `j % weight.shape[2]` logic. If you add additional history dimensions, perhaps it makes sense to set those weights to zero, so that the model doesn't use them at initialisation? In some way that makes sense, because that is how the model was trained before. Are you worried that this initialisation might make it hard for the model to learn to use the new history dimensions?

scottcha replied:
In my initial adhoc testing I was under the impression that copying the previous weights was better than just 0s but based on your comment I decided to test if further.  Using 30 seperate batches from my validation set (default model vars and tuning on the small model/no fine tuning) the mse loss was averaged across all vars was about 20% better using 0s to fill.  So I've gone ahead and updated it filling with zeros.  Good catch!

wesselb replied:
Thanks for the thorough investigation. That's really helpful!

wesselb commented on aurora/model/aurora.py:
```diff
@@ -342,4 +376,6 @@ def configure_activation_checkpointing(self):
     encoder_depths=(6, 8, 8),
     decoder_depths=(8, 8, 6),
+    # One particular static variable requires a different normalisation.
+    surf_stats={"z": (-3.270407e03, 6.540335e04)},
```
This was removed on `main`. Sorry for the hassle!

scottcha replied:
Done.

wesselb reviewed (commented):
@scottcha Amazing work. I've got a few very minor comments, and then I think this is ready to go once CI passes!

Thank you!!

scottcha commented:
Ok, pushed changes to address the feedback.  Thanks!

wesselb commented:
@scottcha I think this is ready to be merged. Would you be able to run the formatter over the PR once? I think that should fix the formatting check.

The easier way by doing this is by setting up `pre-commit` and running that on all files. That should perfectly match the formatter ran in the checks.

scottcha commented:
Got it, handn't noticed the pre-commit component.  Pushed changes to resolve those issues.
thanks!

wesselb reviewed (approved):
Alright, this looks ready to go! Thanks, @scottcha :)
